File: BinaryTreeRightSideView.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def rightSideView(self, root):
        if not root:
            return []
        curr = [root]
        res = []
        while curr:
            res.append(curr[-1].val)
            nxt = []
            for node in curr:
                if node.left:
                    nxt.append(node.left)
                if node.right:
                    nxt.append(node.right)
            curr = nxt
        return res

File: CPP/Second/binary_tree_right_side_view.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> res;
        vector<TreeNode*> curr, next;
        if (!root)
            return res;
        curr.push_back(root);
        while(!curr.empty()) {
            res.push_back(curr.back()->val);
            next.clear();
            for (auto node : curr) {
                if (node->left)
                    next.push_back(node->left);
                if (node->right)
                    next.push_back(node->right);
            }
            curr = next;
        }
        return res;
    }
};


